cpp/src/groupby/sort/group_single_pass_reduction_util.cuh

```diff
@@ -21,20 +21,133 @@
 namespace groupby {
 namespace detail {
+
+// ArgMin binary operator with tuple of (value, index)
+template <typename T>
+struct ArgMin {
+  CUDA_HOST_DEVICE_CALLABLE auto operator()(thrust::tuple<T, size_type> const& lhs,
+                                            thrust::tuple<T, size_type> const& rhs) const
+  {
+    if (thrust::get<1>(lhs) == cudf::detail::ARGMIN_SENTINEL)
+      return rhs;
+    else if (thrust::get<1>(rhs) == cudf::detail::ARGMIN_SENTINEL)
+      return lhs;
+    else
+      return thrust::get<0>(lhs) < thrust::get<0>(rhs) ? lhs : rhs;
+  }
+};
+
+// ArgMax binary operator with tuple of (value, index)
+template <typename T>
+struct ArgMax {
+  CUDA_HOST_DEVICE_CALLABLE auto operator()(thrust::tuple<T, size_type> const& lhs,
+                                            thrust::tuple<T, size_type> const& rhs) const
+  {
+    if (thrust::get<1>(lhs) == cudf::detail::ARGMIN_SENTINEL)
+      return rhs;
+    else if (thrust::get<1>(rhs) == cudf::detail::ARGMIN_SENTINEL)
+      return lhs;
+    else
+      return thrust::get<0>(lhs) > thrust::get<0>(rhs) ? lhs : rhs;
+  }
+};
+
+struct get_tuple_second_element {
+  template <typename T>
+  __device__ size_type operator()(thrust::tuple<T, size_type> const& rhs) const
+  {
+    return thrust::get<1>(rhs);
+  }
+};
+
+/**
+ * @brief Functor to store the boolean value to null mask.
+ */
+struct bool_to_nullmask {
+  mutable_column_device_view d_result;
+  __device__ void operator()(size_type i, bool rhs)
+  {
+    if (rhs) {
+      d_result.set_valid(i);
+    } else {
+      d_result.set_null(i);
+    }
+  }
+};
+
+/**
+ * @brief Returns index for non-null element, and SENTINEL for null element in a column.
+ *
+ */
+struct null_as_sentinel {
+  column_device_view const col;
+  size_type const SENTINEL;
+  __device__ size_type operator()(size_type i) const { return col.is_null(i) ? SENTINEL : i; }
+};
+
+/**
+ * @brief Value accessor for column which supports dictionary column too.
+ *
+ * @tparam T Type of the underlying column. For dictionary column, type of the key column.
+ */
+template <typename T>
+struct value_accessor {
+  column_device_view const col;
+  bool const is_dict;
+  value_accessor(column_device_view const& col) : col(col), is_dict(cudf::is_dictionary(col.type()))
+  {
+  }
+
+  __device__ T value(size_type i) const
+  {
+    if (is_dict) {
+      auto keys = col.child(dictionary_column_view::keys_column_index);
+      return keys.element<T>(static_cast<size_type>(col.element<dictionary32>(i)));
+    } else {
+      return col.element<T>(i);
+    }
+  }
+  CUDA_HOST_DEVICE_CALLABLE auto operator()(size_type i) const { return value(i); }
+};
+
+/**
+ * @brief Null replaced value accessor for column which supports dictionary column too.
+ * For null value, returns null `init` value
+ *
+ * @tparam T Type of the underlying column. For dictionary column, type of the key column.
+ */
+template <typename T>
+struct null_replaced_value_accessor : value_accessor<T> {
```

Comment:
Same question here. Does the existing `null_replacement_iterator` not work?

Reply:
`null_replacement_iterator` right now doesn't support dictionary columns. This functor does. If `null_replacement_iterator` is updated to support dictionary too, it will add to all kernels using it.
Can I add dictionary support to `null_replacement_iterator<T>` (T is underlying type, not dictionary32 for dictionary type)?

Reply:
@davidwendt  thoughts?

Reply:
Can you use this?
https://github.com/rapidsai/cudf/blob/8b02ca32c8f9004de5bbb1d76bd77b17358538b4/cpp/include/cudf/dictionary/detail/iterator.cuh#L110-L112

Reply:
It would be better to use the indices for any cudf operations where possible for both run-time and compile-time performance.

Reply:
This would be useful for argmin, argmax, which already uses dictionary32 type (which is wrapped indices)
`null_replaced_value_accessor` is for accessing values for SUM, PROD, MIN, MAX which needs values. I don't see how `get_indices_annotated()` would be useful to replace `null_replaced_value_accessor`.

Reply:
Yes, that makes sense. I did not actually go through the PR. I was just basing off the title. A sort-groupby would only need the indices to create the groups. You certainly would need the actual values to do SUM, PROD.  But I think it should be possible to do MIN, MAX with just the indices since the index values represent the sorted (ascending) keys. That is, the lowest non-null index value will point to the MAX and the largest non-null index value would point to the MAX. So basically doing ARGMIN or ARGMAX first and then just retrieving the actual key values at the end.

Reply:
This raises a question. 
what output column, should _groupby min/max_ for **dictionary column of a base type** produce ?
  the base type column OR dictionary column of the base type?

Reply:
hash groupby produces base type column as output.
If we use gather with ARGMIN, or ARGMAX for MIN, or MAX, it would create dictionary column. (added one more test for this, and updated sort groupby to fix this)

Reply:
This sounds correct to me. Aggregates like min/max return values that already exist in the column so the output would have the same keys as the input. Whereas, sum/prod create totally new values.

Reply:
Also, here is an example using the the dictionary-pair-iterator along with a null-replacement transformer.
https://github.com/rapidsai/cudf/blob/f0fa255add77daf6fd14b714286d01d5c2b4d082/cpp/src/reductions/simple.cuh#L142-L146
I'm inclined to prefer your approach here instead since it simplifies the caller to one value-accessor. The only thing that makes me nervous is that `col.element<dictionary32>(i)` would be included/inlined for every type and that function contains it's own type-dispatcher call in it. But technically every type is potentially a dictionary key type so I think the same amount of code is generated either way. Anyway, it may be worth looking into using this null-replacement accessor in the reductions code too.